cpu/atmega256rfr2/periph/hwrng.c

```diff
@@ -0,0 +1,151 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_atmega256rfr2
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the CPU initialization
+ *
+ * @author      Josua Arndt <jarndt@ias.rwth-aachen.de>
+ * @}
+ */
+
+#include "cpu.h"
+
+#include "periph_conf.h"
+#include "periph/hwrng.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void transition(void){
+    /* Make sure transceiver is finished with transition */
+    while( (TRX_STATUS&0x1f)==0x1f){}
+}
+
+void hwrng_init(void)
+{
+    /* Nothing to do here. */
+}
+
+/*
+ * Random Number Generator
+ *
+ * The radio transceiver incorporates a 2-bit, noise observing, true random number
+ * generator to be used to:
+ *      Generate random seeds for CSMA-CA algorithm
+ *      (see"Extended Operating Mode" on page 47)
+ *
+ *      Generate random values for AES key generation
+ *      (see "Security Module (AES)" on page 99);
+ *
+ * The values are stored in bits RND_VALUE of register PHY_RSSI. The random number
+ * is updated every tTR29 = 1 μs in Basic Operation Mode receive states with locked PLL.
+ *
+ * Note, if the PLL is not locked or unlocks in receive states or either antenna
+ * diversity or RPC mode is enabled, the RND_VALUE is zero.
+ */
+void hwrng_read(void *buf, unsigned int num)
+{
+    DEBUG("Hardware random bytes to get: %d\n", num);
+
+    /* Make sure transceiver is not in transition */
+    transition();
+
+    uint8_t prevRxSync = RX_SYN;
+    uint8_t prevTrxRpc = TRX_RPC;
+    uint8_t prevState = TRX_STATE;
+    uint8_t prevTrxPr = TRXPR;
+    uint8_t prevAntDiv = ANT_DIV;
+
+    xtimer_ticks32_t wait= xtimer_ticks_from_usec(1);
+
+    /* Manual p. 119 RX_PDT_DIS (Register RX_SYN) is set to 0 */
+    RX_SYN &= ~(1 << RX_PDT_DIS);
+
+    /* Manual p. 105 RPC/SRT has to be disabled. */
+    TRX_RPC &= ~(1<<RX_RPC_CTRL1 | 1<<RX_RPC_CTRL0);
+    /* randome ready time 2*1us + 24us (pll Ready) */
+
+    /* Disable antenna diversity */
+    ANT_DIV &= ~(1<<ANT_DIV_EN);
+
+    /* Make sure transceiver is not in sleep */
+    if (TRXPR & (1<<SLPTR)) {
+        /* wake up the transceiver */
+        TRXPR &= ~(1<<SLPTR);
+    }
+
+    transition();
+
+    /* Manual p. 119 radio transceiver has to be in Basic Operating Mode receive state*/
+    if ( (TRX_STATUS&0x1f)!= 0x06) {
+        /* can not be forced to PLL_ON from TRX_OFF */
+        if( (TRX_STATUS&0x1f)!= 0x08 ){
+            /* Set to PLL_ON */
+            TRX_STATE = 0x09;
+        }else{
+            /* Force to PLL_ON */
+            TRX_STATE = 0x04;
+        }
+
+        transition();
+
+        /* Set to RX_ON */
+        TRX_STATE = 0x06;
+
+        transition();
+    }
+
+    for (size_t byteCount = 0; byteCount < num; ++byteCount) {
+        uint8_t rnd = 0;
+        for (uint8_t i = 0; i < 4; ++i) {
+            /* Random value update period 1us p. 47*/
+            xtimer_spin(wait);
```

Comment:
Is this necessary?
I imagine that the CPU instructions leading up to this point take at least 1 µs on atmegas?